### PR TITLE
[Bounty] Deepr PyHealth 2.0 ver.

### PR DESCRIPTION
**Pull request by joshuasteier**
## Contributor Information
- **Name:** Joshua Steier
- **Contribution Type:** Model

## Description
Updated the Deepr model to align with PyHealth 2.0 conventions:
- Migrated from `BaseEHRDataset` to `SampleDataset`
- Replaced manual tokenization with `EmbeddingModel`
- Updated label handling to use `get_output_size()` and direct tensor access
- Added learnable gap embedding for nested sequence flattening
- Supports both 3D and 4D input tensors

## Files to Review
- `pyhealth/models/deepr.py`
- `tests/core/test_deepr.py`

## Testing
```bash
python -m pytest -q tests/core/test_deepr.py -s
# Result: 5 passed
```

### CODE REVIEW

**Comment by Logiquo** [comment]
Could you also provide a example notebook in the `examples` folder to demonstrate how to use this model? thank you.

File: pyhealth/models/deepr.py [hunk 1]
```diff
@@ -1,46 +1,36 @@
...
-    Examples:
-        >>> from pyhealth.models import DeeprLayer
-        >>> input = torch.randn(3, 128, 5)  # [batch size, sequence len, input_size]
-        >>> layer = DeeprLayer(5, window=4, hidden_size=7) # window does not impact the output shape
-        >>> outputs = layer(input)
-        >>> outputs.shape
-        torch.Size([3, 7])
+        feature_size: Input feature size (embedding_dim).
+        window: Sliding window radius d; kernel_size = 2*d + 1.
+        hidden_size: Number of convolution filters.
```
Comment (Logiquo) [inline pyhealth/models/deepr.py:21]: Could you update the usage examples in the docstring similar to the old docstring? This could help us to generate a nice document.

File: pyhealth/models/deepr.py [hunk 2]
```diff
@@ -49,244 +39,166 @@ def forward(
...
-                        [0.2665]], device='cuda:0', grad_fn=<SigmoidBackward0>),
-            'y_true': tensor([[1.],
-                        [0.]], device='cuda:0'),
-            'logit': tensor([[-1.2110],
-                        [-1.0126]], device='cuda:0', grad_fn=<AddmmBackward0>)
-        }
+        dataset: SampleDataset with fitted input and output processors.
+        embedding_dim: Size of the intermediate embedding space.
+        hidden_dim: Number of convolution filters produced by DeeprLayer.
+        window: Sliding window radius d (kernel_size = 2*d + 1).
```
Comment (Logiquo) [inline pyhealth/models/deepr.py]: Could you update the usage examples in the docstring similar to the old docstring? This could help us to generate a nice document.

File: pyhealth/models/deepr.py [hunk 3]
```diff
@@ -49,244 +39,166 @@ def forward(
...
+        if isinstance(value, torch.Tensor):
+            return value
+        return torch.as_tensor(value)
+
+    def _flatten_nested_with_gap(
+        self, x: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Flattens [B, V, T, C] -> [B, L, C] and inserts a gap token between visits."""
+        if x.dim() != 4:
+            raise ValueError(f"Expected 4-D x [B, V, T, C], got {x.shape}")
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/deepr.py]: Fixed

File: pyhealth/models/deepr.py [hunk 4]
```diff
@@ -49,244 +39,166 @@ def forward(
...
-            c: a Tensor of shape [batch size, hidden_size] representing the
-                summarized vector.
+            Tensor of shape [batch size, hidden size].
         """
+        if x.dim() != 3:
+            raise ValueError(f"Expected 3-D x [B, T, C], got {x.shape}")
+
         if mask is not None:
+            if mask.dim() != 2:
+                raise ValueError(f"Expected 2-D mask [B, T], got {mask.shape}")
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/deepr.py]: Fixed

File: pyhealth/models/deepr.py [hunk 5]
```diff
@@ -49,244 +39,166 @@ def forward(
...
+
+    def _prepare_sequence_and_mask(
+        self, x: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        if x.dim() == 3:
+            mask = (x.abs().sum(dim=-1) != 0).to(dtype=x.dtype)  # [B, T]
+            return x, mask
+        if x.dim() == 4:
+            return self._flatten_nested_with_gap(x)
+        raise ValueError(f"Deepr expects embedded features to be 3-D or 4-D, got {x.shape}")
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/deepr.py]: Fixed

File: pyhealth/models/deepr.py [hunk 6]
```diff
@@ -1,46 +1,36 @@
...
-        window: int = 1,
-        hidden_size: int = 3,
-    ):
+    def __init__(self, feature_size: int, window: int = 1, hidden_size: int = 128):
         super(DeeprLayer, self).__init__()
-
-        self.conv = torch.nn.Conv1d(
-            feature_size, hidden_size, kernel_size=2 * window + 1
+        if window < 0:
+            raise ValueError("window must be non-negative")
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/deepr.py]: Fixed

File: tests/core/test_deepr.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+        with torch.no_grad():
+            ret = model(**data_batch)
+
+        self.assertIn("loss", ret)
+        self.assertIn("y_prob", ret)
+
+
+if __name__ == "__main__":
+    unittest.main()
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply tests/core/test_deepr.py:169]: Fixed

File: pyhealth/models/deepr.py [hunk 8]
```diff
@@ -1,46 +1,36 @@
...
-        hidden_size: number of conv filters (motif size, p, in the original paper)
-    Examples:
-        >>> from pyhealth.models import DeeprLayer
-        >>> input = torch.randn(3, 128, 5)  # [batch size, sequence len, input_size]
-        >>> layer = DeeprLayer(5, window=4, hidden_size=7) # window does not impact the output shape
-        >>> outputs = layer(input)
-        >>> outputs.shape
-        torch.Size([3, 7])
+        feature_size: Input feature size (embedding_dim).
+        window: Sliding window radius d; kernel_size = 2*d + 1.
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/deepr.py]: Fixed

File: pyhealth/models/deepr.py [hunk 9]
```diff
@@ -49,244 +39,166 @@ def forward(
...
-            'y_true': tensor([[1.],
-                        [0.]], device='cuda:0'),
-            'logit': tensor([[-1.2110],
-                        [-1.0126]], device='cuda:0', grad_fn=<AddmmBackward0>)
-        }
+        dataset: SampleDataset with fitted input and output processors.
+        embedding_dim: Size of the intermediate embedding space.
+        hidden_dim: Number of convolution filters produced by DeeprLayer.
+        window: Sliding window radius d (kernel_size = 2*d + 1).
     """
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/deepr.py:113]: Fixed

File: pyhealth/models/deepr.py [hunk 10]
```diff
@@ -49,244 +39,166 @@ def forward(
...
+            x: Tensor of shape [batch size, seq len, feature size].
+            mask: Optional tensor of shape [batch size, seq len], 1 valid / 0 pad.
+
         Returns:
-            c: a Tensor of shape [batch size, hidden_size] representing the
-                summarized vector.
+            Tensor of shape [batch size, hidden size].
         """
+        if x.dim() != 3:
+            raise ValueError(f"Expected 3-D x [B, T, C], got {x.shape}")
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/deepr.py]: Fixed

File: pyhealth/models/deepr.py (hunk 1 above)
Comment (joshuasteier) [reply pyhealth/models/deepr.py:21]: Thank you, should be fixed now.

File: pyhealth/models/deepr.py (hunk 2 above)
Comment (joshuasteier) [reply pyhealth/models/deepr.py]: Thank you, should be fixed now

**Comment by joshuasteier** [comment]
@jhnwu3, @LogicFan, I added an example notebook in addition to the changes above, thank you.
